kaitaistruct.py

```diff
@@ -38,14 +38,32 @@ def from_file(cls, filename):
             f.close()
             raise
 
+    @classmethod
+    def to_file(cls, filename, data=None):
+        f = open(filename, 'rb')
+        try:
+            return cls(KaitaiStream(f), _mode='w', _data=data)
```

Comment:
What are the `_mode` and `_data` parameters supposed to do? They aren't accepted by `KaitaiStruct.__init__` or by the `__init__`s generated by the compiler.

Reply:
`__init__`s generated by the compiler (or manually currently)

Reply:
Here is use-case of what it could look like, when implemented on the compiler side...

Generated class:
```python
class UnityStrDat(KaitaiStruct):
    def __init__(self, _io, _parent=None, _root=None, _mode="r", _data=None):
        self._io = _io
        self._parent = _parent
        self._root = _root if _root else self
        if _mode == "r": self._read()
        elif _mode == "w" and _data: self._write(_data)

    def _read(self):
        self.header = UnityStrDat.HeaderRaw(self._io, self, self._root)
        self.num_strings = self._io.read_u4le()
        self.strings_arr = [None] * (self.num_strings)
        for i in range(self.num_strings):
            self.strings_arr[i] = UnityStrDat.StringStruct(self._io, self, self._root)

    def _write(self, data):
        if not data: return
        UnityStrDat.HeaderRaw(self._io, self, self._root, _mode="w", _data=data[0])
        self._io.write_u4le(len(data[1]) - 1)
        for i in range(len(data[1])):
            UnityStrDat.StringStruct(self._io, self, self._root, _mode="w", _data=data[1][i][1])


    class HeaderRaw(KaitaiStruct):
        def __init__(self, _io, _parent=None, _root=None, _mode="r", _data=None):
            self._io = _io
            self._parent = _parent
            self._root = _root if _root else self
            if _mode == "r": self._read()
            elif _mode == "w" and _data: self._write(_data)

        def _read(self):
            self.header_raw_bytes = self._io.read_bytes(500)

        def _write(self, data=None):
            if not data and self.header_raw_bytes: data = self.header_raw_bytes
            l = len(data)
            if not data or 500 != l:
                raise Exception("No header provided or wrong size (%d, instead of %d)" % (l, self._length))
            self._io.write_bytes(data)

    class StringStruct(KaitaiStruct):
        def __init__(self, _io, _parent=None, _root=None, _mode="r", _data=None):
            self._io = _io
            self._parent = _parent
            self._root = _root if _root else self
            if _mode == "r": self._read()
            elif _mode == "w" and _data: self._write(_data)

        def _read(self):
            self.s_size = self._io.read_u4le()
            self.string = self._io.read_bytes(self.s_size, align=4).decode(u"utf-8")

        def _write(self, data):
            if not data: return
            data = data.encode(self._encoding)
            self._io.write_u4le(len(data))
            self._io.write_bytes(data, align=4)
```
With usage like this:
```python

    if app_args.pack:
        a = UnityStrDat.from_file(file_name)
        a.close()
        data = load_old(os.path.splitext(file_name)[0]+'_data.old')
        new_name = "os.path.splitext(file_name)[0] + "_fixed" + os.path.splitext(file_name)[1]
        path_out = os.path.dirname(new_name)
        if path_out != '' and not os.path.exists(path_out):
            os.makedirs(path_out, exist_ok=True)
        a = UnityStrDat.to_file(new_name, (a.header1.header_raw_bytes, data))
        a.close()
    else:
        data = []
        with UnityStrDat.from_file(file_name) as a:
            for i in a.strings_arr:
                data.append([i.string, ''])

Reply:
Separate modes for reading and writing is IMHO a bad idea. It would require reopening the format. In the serialization proposal it was proposed to not to serialize from scratch, but cache the changes in memory and then apply them. So, if we ever have modes, there will be only 2, read-only and read-write. If we have 2 modes, they can be toggled by a boolean variable.

Reply:
If you can implement cached changes I am all for it.

There can be more modes possible as underlying streams can support more, like appending mode of file IO stream so it's good idea not to shoot ourselves in the leg and not to restrict modes to just boolean.

Reply:
>There can be more modes possible as underlying streams can support more, like appending mode of file IO stream so it's good idea not to shoot ourselves in the leg by restricting modes to just boolean.

OK, append-only mode is needed too, because of FS_APPEND_FL.